src/dense_vector.rs

```diff
@@ -0,0 +1,232 @@
+use crate::Ix1;
+use ndarray::{self, ArrayBase};
+use num_traits::identities::Zero;
+
+/// A trait for types representing dense vectors, useful for expressing
+/// algorithms such as sparse-dense dot product, or linear solves.
+pub trait DenseVector {
+    type Owned;
+    type Scalar;
+
+    /// The dimension of the vector
+    fn dim(&self) -> usize;
+
+    /// Random access to an element in the vector.
+    ///
+    /// # Panics
+    ///
+    /// If the index is out of bounds
+    fn index(&self, idx: usize) -> &Self::Scalar;
+
+    /// Create an owned version of this dense vector type, filled with zeros
+    fn zeros(dim: usize) -> Self::Owned;
```

Comment:
Are you thinking `zeros` will always be the same size as `self` here?

Reply:
There's no `self` here, this one is static, I think I'll probably need it for some implementations (like `to_dense`).

Reply:
I was testing how the trait would interact with const generics. `Self::Owned` can not be constrained to a generic size `[N; M], and would need to be a `Vec` or equivalent.

Reply:
Not that this is a blocker in any way

Reply:
Right, I was thinking a bit about arrays, but I don't have a plan for them right now.

Reply:
I doubt you'll want to use a sparse library for arrays anyway, those kind of problems are probably a better fit for dense libraries which could use the extra information more efficiently